src/include/statistics/stats_aggregator.h

```diff
@@ -22,148 +22,44 @@
-  StatsAggregator &operator=(const StatsAggregator &) = delete;
-  StatsAggregator(StatsAggregator &&) = delete;
-  StatsAggregator &operator=(StatsAggregator &&) = delete;
+  StatsAggregator(int64_t aggregation_interval)
+      : aggregation_interval_ms_(aggregation_interval) {}
 
-  StatsAggregator(int64_t aggregation_interval_ms);
-  ~StatsAggregator();
+  void Terminate() override;
 
-  //===--------------------------------------------------------------------===//
-  // ACCESSORS
-  //===--------------------------------------------------------------------===//
+  void RunTask() override;
 
-  // Global singleton
-  static StatsAggregator &GetInstance(
-      int64_t aggregation_interval_ms = STATS_AGGREGATION_INTERVAL_MS);
-
-  // Get the aggregated stats history of all exited threads
-  inline BackendStatsContext &GetStatsHistory() { return stats_history_; }
-
-  // Get the current aggregated stats of all threads (including history)
-  inline BackendStatsContext &GetAggregatedStats() { return aggregated_stats_; }
-
-  //===--------------------------------------------------------------------===//
-  // HELPER FUNCTIONS
-  //===--------------------------------------------------------------------===//
-
-  // Register the BackendStatsContext of a worker thread to global Stats
-  // Aggregator
-  void RegisterContext(std::thread::id id_, BackendStatsContext *context_);
-
-  // Unregister a BackendStatsContext. Currently we directly reuse the thread id
-  // instead of explicitly unregistering it.
-  void UnregisterContext(std::thread::id id);
-
-  // Aggregate the stats of current living threads
-  void Aggregate(int64_t &interval_cnt, double &alpha,
-                 double &weighted_avg_throughput);
-
-  // Launch aggregator thread
-  void LaunchAggregator();
-
-  // Terminate aggregator thread
-  void ShutdownAggregator();
+  /**
+   * Aggregate metrics from all threads which have collected stats,
+   * combine with what was previously in catalog
+   * and insert new total into catalog
+   */
+  void Aggregate();
 
  private:
-  //===--------------------------------------------------------------------===//
-  // MEMBERS
-  //===--------------------------------------------------------------------===//
-
-  // Stores stats of exited threads
-  BackendStatsContext stats_history_;
-
-  // Stores all aggregated stats
-  BackendStatsContext aggregated_stats_;
-
-  // Protect register and unregister of BackendStatsContext*
-  std::mutex stats_mutex_{};
-
-  // Map the thread id to the pointer of its BackendStatsContext
-  std::unordered_map<std::thread::id, BackendStatsContext *> backend_stats_{};
-
-  // How often to aggregate all worker thread stats
   int64_t aggregation_interval_ms_;
-
-  // Number of threads registered
-  int thread_number_;
-
-  int64_t total_prev_txn_committed_;
-
-  // Stats aggregator background thread
-  std::thread aggregator_thread_;
-
-  // CV to signal aggregator if finished
+  std::mutex mutex_;
```

Comment:
Although everyone (read Andy) hates mutexes, I do think this is the right way to go about it.

-- Priyatham, Siva and Vamshi